### ISSUE TITLE
Improve material model evaluation to respect averaging and formulation

### ISSUE THREAD

**Issue by gassmoeller** [geodynamics/aspect#3358]
In #3335 we found another place where the material model evaluation did not respect the formulation because the calling code forgot to replace the density by the reference density. I suspect there are more cases like this hidden in some plugins. The problem is that this is something that needs to be done nearly always (except for the Stokes RHS), but has to be handled outside of the material model at the moment. Should this be something the material model does internally? The same question could be asked for the averaging (which is sometimes done, sometimes not), and for the check for negative material properties.

I think this should be handled inside the material model and I would imagine a function like this:
```
MaterialModel::Interface<dim>::safely_evaluate(const MaterialModel::MaterialModelInputs<dim> &in,
                                               MaterialModel::MaterialModelOutputs<dim> &out))
```

The implementation of this function (in the interface class, not every individual plugin) would then call the existing `evaluate` function and afterwards do the averaging and substitution of density (if required by the formulation). Finally it could also check for negative material properties (although we should benchmark first how much the AssertThrow would cost).

I am not sure about the name of this function I did not come up with a better one though. Maybe `evaluate_effective_properties`?

**Comment by tjhei** [comment]
> Should this be something the material model does internally?

I think there are two separate issues that might require different approaches:
1. We sometimes need to replace the density (but depending on the formulation and what equation we work on). So I don't see how this can go into safely_evaluate as this function needs to know why I want the density right now. Maybe we need introduce two different densities to the material outputs (they could be filled by some helper function).
2. For averaging: yes, I also noticed the other day that we are not averaging material outputs in visualization (which I think would be sensible to do). Generally, we want to always average, which means it should be in some function (helper that calls evaluate).

> and for the check for negative material properties.

This reminds me of #2025 :-)

**Comment by tjhei** [comment]
So my suggestion:
1. Implement a function safe_evaluation (maybe even outside of MaterialModel) and use it everywhere in ASPECT.
2. add new outputs ``temperature_density`` (or some different name) that gets filled in 1. based on the temperature_formulation and use this in the assemblers as needed (review all usages of density)
3. Make the new function do evaluate() and average() and optionally check values (are all values filled, densities positive, etc.)

**Comment by gassmoeller** [comment]
I agree with 1 and 3, but I have some concerns about 2. We had functions that produced `reference_density` for every material model in the past and worked hard to get rid of them, because it was unclear which places used them and for what purpose. Additionally, they introduce some uncertainties for implementing plugins or calling material models. E.g when implementing a new material model, do I need to fill `temperature_density`? (no, but all other properties of MaterialModelOutputs). When I evaluate a material model, can I safely use `temperature_density` afterwards? (no, only if you called `safely_evaluate`, but not if you called `evaluate`). 

What about we add a parameter to `safely_evaluate` that is called `respect_formulation`? We could set the default value to true, and switch it off when assembling the Stokes equation. So by default the density would be replaced by the reference density, except for the cases we do not want it do be replaced.

**Comment by tjhei** [comment]
> that is called `respect_formulation`

I agree with your points, but it is not as simple as "respect formulation = true/false": we also need to say "this is for the Stokes system" or "this is for the advection/temperature system".
Alternatively, we could always generate a separate ``density_for_temperature_based_on_current_formulation`` in the new ``safely_evaluate`` function (and not store it inside the material outputs).

**Comment by gassmoeller** [comment]
I have thought about this a bit more and I think you are right, we need an additional output (either as separate variable or inside MaterialModelOutputs, maybe as `AdditionalMaterialOutputs` that is created only if necessary to hide it from the users?). However, if we do it exactly like you proposed the burden would still fall on the calling side to use `density_for_temperature_based_on_current_formulation` instead of `density`, so we would need to change all of our plugins and hope that every user-plugin would do the same.
I think it might be better to reverse the roles that you suggested. The vast majority of current applications *should* use the reference density for BA/ALA and full density for ICA/PDA, but only some do the necessary replacement. As far as I see the only places that should use the full density for BA/ALA are the assembly for the Stokes right hand side, the density mesh refinement, and maybe the density visualization postprocessor (although even that could be discussed). So I would suggest to make the reference density the default and do the following in safely_evaluate:

```
if (formulation == reference_density && adiabatic_conditions.is_initialized() && out.additional_outputs.full/dynamics_densities exists)
 {
  out.additional_outputs.full/dynamic_densities = out.densities;
  out.densities = adiabatic_conditions.densities;
 }
```

The one thing that would not work with this approach at the moment is that we allow weird cases like: use reference density profile for Stokes, but full density for temperature; and: use reference density for temperature but full density (ICA/PDA) for Stokes. Maybe we should just disallow these combinations and enforce that if one equation uses a reference density, all of them should.
What do you think?

**Comment by tjhei** [comment]
> I think it might be better to reverse the roles that you suggested. The vast majority of current applications _should_ use the reference density for BA/ALA and full density for ICA/PDA, but only some do the necessary replacement.

That is not a bad solution. One disadvantage is that out.density no longer contains what the user puts into it. This might be confusing (what do you want to visualize/postprocess for example?).

Another option would be to introduce two new variables: ``dynamic_density`` and ``adiabatic_density`` and remove all usages of ``density`` in the code. We could even rename the material model output to ``dynamic_density`` and deprecate the old name. Not sure if this is a good idea. :-)

**Comment by gassmoeller** [comment]
> One disadvantage is that out.density no longer contains what the user puts into it. This might be confusing

Yes I can see that being confusing.

> Another option would be to introduce two new variables: dynamic_density and adiabatic_density and remove all usages of density in the code

Then we are back to the current state where the calling side has to decide which density to use based on the formulation (or alternatively, if `adiabatic_density` has different content depending on the formulation the name would be confusing, because it is not the adiabatic_density for ICA/PDA). Maybe this discussion just shows that there is no better way than what we have at the moment (let the calling side decide if to use densities or adiabatic_conditions.density). I would be willing to drop this point for now, but I think we agree 1. and 3. of your second post should be inside safely_evaluate.
Should I make a first version of that and we see how it looks? It might have to wait a week though.
